### PR TITLE
[FEATURE] Api Wrapper to handle connection issues

### PR DESCRIPTION
**Pull request by DayBr3ak**
Short Description: 

I'm trying to handle connections issues with this wrapper.
I've been testing it for an hour for now. Maybe you want to review/test this yourself.
For now it just does some retries (5) on the request if it failed.
And throw an error if it fails instead of an error like `__getitem__ on bool ` ... etc

I'll update this when I've done enough testing

### CODE REVIEW

File: pokemongo_bot/api_wrapper.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        try_cnt = 0
+        while True:
+            self._api._req_method_list = [req_method for req_method in api_req_method_list] # api internally clear this field after a call
+            result = self._api.call()
+            if result is None:
+                try_cnt += 1
+                logger.log('Server seems to be busy or offline - try again - {}/{}'.format(try_cnt, max_retry), 'red')
+                if try_cnt >= max_retry:
+                    raise ServerBusyOrOfflineException()
+                sleep(10)
```
Comment (cwild) [inline pokemongo_bot/api_wrapper.py]: Where are you importing sleep from?

Comment (DayBr3ak) [inline pokemongo_bot/api_wrapper.py]: good catch

File: pokemongo_bot/api_wrapper.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        if not self._can_call():
+            return False
+
+        api_req_method_list = self._api._req_method_list
+        result = None
+        max_retry = 5 # TODO configuration param
+        try_cnt = 0
+        while True:
+            self._api._req_method_list = [req_method for req_method in api_req_method_list] # api internally clear this field after a call
+            result = self._api.call()
```
Comment (cwild) [inline pokemongo_bot/api_wrapper.py]: Worth wrapping in a ```try...except``` block perhaps? Exceptions can be raised from here still.

Comment (DayBr3ak) [inline pokemongo_bot/api_wrapper.py]: What else can be raised here if I may ask?

File: pokemongo_bot/api_wrapper.py (hunk 1 above)
Comment (douglascamata) [inline pokemongo_bot/api_wrapper.py]: holy shit, please do not wait 10 seconds... 1 second is good enough.

Comment (DayBr3ak) [inline pokemongo_bot/api_wrapper.py]: done

File: pokemongo_bot/api_wrapper.py (hunk 2 above)
Comment (cwild) [inline pokemongo_bot/api_wrapper.py]: Well I am not familiar with the underlying API. I'd say whenever you call a function, there's a risk of an Exception being raised (maybe not even in the current API release). So it's a matter of deciding if you want to harden your API wrapper to handle those errors or not.

Any of the core maintainers care to share their opinions?

Comment (DayBr3ak) [inline pokemongo_bot/api_wrapper.py]: You're right, api.call use request.request so I could check for those exceptions

**Comment by douglascamata** [comment]
:+1:

**Comment by TheSavior** [comment]
I wonder if this kind of retry logic should be moved upstream. @tejado, do you have any thoughts?

**Comment by TheSavior** [comment]
👍 

[Approved with PullApprove](https://pullapprove.com/PokemonGoF/PokemonGo-Bot/pull-request/1459/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=PokemonGoF/PokemonGo-Bot)

**Comment by tejado** [comment]
Yes, this would be a fantastic idea. Already had similar thoughts. I will implement it also in pgoapi 1.2.0

**Comment by douglascamata** [comment]
While @tejado doesn't release `pgoapi` 1.2.0 we keep this wrapper.

**Comment by douglascamata** [comment]
👍

**Pull request merged**